cores/nRF5/HardwarePWM.cpp

```diff
@@ -350,3 +278,72 @@ uint8_t HardwarePWM::freeChannelCount(void) const
   return MAX_CHANNELS - usedChannelCount();
 }
 
+// returns true ONLY when (1) no PWM channel has a pin, and (2) the owner token is nullptr
+bool HardwarePWM::takeOwnership(uintptr_t token)
+{
+  if (token == 0) {
+    LOG_LV1("HwPWM", "zero / nullptr is not a valid ownership token (attempted use in takeOwnership)");
+    return false;
+  }
+
+  if (token == this->_owner_token) {
+    LOG_LV1("HwPWM", "failing to acquire ownership because already owned by requesting token (cannot take ownership twice)");
+    return false;
+  }
+
+  if (this->_owner_token != 0) return false;
+  if (this->usedChannelCount() != 0) return false;
+  if (this->enabled()) return false;
+
+  if (isInISR())
+  {
+    UBaseType_t intr_status = taskENTER_CRITICAL_FROM_ISR();
+    _owner_token = token;
+    taskEXIT_CRITICAL_FROM_ISR(intr_status);
+  }else
+  {
+    taskENTER_CRITICAL();
+    _owner_token = token;
+    taskEXIT_CRITICAL();
+  }
+
+  return true;
+}
+
+// returns true ONLY when (1) no PWM channel has a pin attached, and (2) the owner token matches
+bool HardwarePWM::releaseOwnership(uintptr_t token)
+{
+  if (token == 0) {
+    LOG_LV1("HwPWM", "zero / nullptr is not a valid ownership token (attempted use in releaseOwnership)");
+    return false;
+  }
+
+  if (!this->isOwner(token)) {
+    LOG_LV1("HwPWM", "attempt to release ownership when not the current owner");
+    return false;
+  }
+
+  if (this->usedChannelCount() != 0) {
+    LOG_LV1("HwPWM", "attempt to release ownership when at least on channel is still connected");
+    return false;
+  }
+
+  if (this->enabled()) {
+    LOG_LV1("HwPWM", "attempt to release ownership when PWM peripheral is still enabled");
+    return false; // if it's enabled, do not allow ownership to be released, even with no pins in use
+  }
+
+  if (isInISR())
+  {
+    UBaseType_t intr_status = taskENTER_CRITICAL_FROM_ISR();
+    _owner_token = 0;
```

Comment:
Bug -- you added a race condition by unconditionally changing the value.

Reply:
The condition won't happen, nothing go through the task critical

Reply:
It's not the task critical.  It's the unconditional setting of the value.

You clearly did not understand the code you were replacing.